Initial OpInfo-based testing for NJTs

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131945
* #131937
* __->__ #131704
* #131898

This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) to eager
    * `test_backward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `BinaryUfuncInfo`s utilize a new `sample_inputs_binary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `ReductionOpInfo`s utilize a new `sample_inputs_njt_reduction()` which covers full reductions, reductions over the jagged dim, and reductions over the non-jagged dim
* Several xfails were added to get things passing

TODO (future PRs):
* Pass non-contiguous / non-contiguous with holes NJTs (maybe we should have separate tests for these? most ops don't support NJTs with holes today)
* Mixed (NT, T), (T, NT) inputs for binary ops
* Handle other types of OpInfos (beyond unary pointwise, binary pointwise, and reduction) by manually by writing sample_inputs_funcs
* Address all xfails via fixes

jbschlosser committed:
OpInfo-based testing for NJTs

[ghstack-poisoned]

jbschlosser committed:
Update on "OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward output to eager
    * `test_backward_compile()` - compares forward output and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
* Several xfails were added for things that are known to be failing

TODO:
* Handle OpInfo translation for non-unary pointwise ops
* Write specific sample inputs funcs for certain ops when necessary
* Finish xfails for backward / compile tests
* Pass non-contiguous / non-contiguous with holes NJTs
* Fix the forward failures!

[ghstack-poisoned]

jbschlosser committed:
Update on "OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward output to eager
    * `test_backward_compile()` - compares forward output and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
* Several xfails were added for things that are known to be failing

TODO:
* Handle OpInfo translation for ops beyond: unary pointwise, binary pointwise, and reduction
* Write specific sample inputs funcs for certain ops when necessary
* Finish xfails for compile tests
* Pass non-contiguous / non-contiguous with holes NJTs
* Fix the forward failures!

[ghstack-poisoned]

jbschlosser committed:
Update on "OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward output to eager
    * `test_backward_compile()` - compares forward output and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
* Several xfails were added for things that are known to be failing

TODO:
* Handle OpInfo translation for ops beyond: unary pointwise, binary pointwise, and reduction
* Write specific sample inputs funcs for certain ops when necessary
* Finish xfails for compile tests
* Pass non-contiguous / non-contiguous with holes NJTs
* Fix the forward failures!

[ghstack-poisoned]

jbschlosser committed:
Update on "OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward output to eager
    * `test_backward_compile()` - compares forward output and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
* Several xfails were added for things that are known to be failing

TODO:
* Handle OpInfo translation for ops beyond: unary pointwise, binary pointwise, and reduction
* Write specific sample inputs funcs for certain ops when necessary
* Finish xfails for compile tests
* Pass non-contiguous / non-contiguous with holes NJTs
* Fix the forward failures!

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -1061,7 +1062,14 @@ def select_int(func, *args, **kwargs):
 
     inp = new_kwargs.pop("input")
-    new_kwargs["dim"] = _wrap_jagged_dim(inp.dim(), new_kwargs["dim"], "select")
+    new_kwargs["dim"] = _wrap_jagged_dim(
+        inp.dim(), new_kwargs["dim"], "select", allow_batch_dim=True
+    )
+
+    # handle batch dim slicing via unbind() for now
+    # TODO: make this more efficient
+    if new_kwargs["dim"] == 0:
+        return inp.unbind()[new_kwargs["index"]]
```
`select(dim=0)` is useful for slicing up other NJTs in the input when computing the unbind-based reference

jbschlosser commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -1437,15 +1442,17 @@ def supported_backward_dtypes(self, device_type):
     def supports_dtype(self, dtype, device_type) -> bool:
         return dtype in self.supported_dtypes(device_type)
 
+    @property
+    def full_name(self):
+        """Returns a full name that helps to uniquely identify this OpInfo."""
+        variant = "." + self.variant_test_name if self.variant_test_name else ""
+        # example: "normal.in_place" where "normal" is the name and "in_place" is the variant
+        return f"{self.name}{variant}"
```
small thing but I really want to specify ops by their full name of `"{name}.{variant}"`. `formatted_name` wasn't quite right because it uses underscores instead of periods

jbschlosser committed:
Update on "Initial OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) to eager
    * `test_backward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `BinaryUfuncInfo`s utilize a new `sample_inputs_binary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
        * **TODO: mixed (NT, T), (T, NT) inputs**
    * `ReductionOpInfo`s utilize a new `sample_inputs_njt_reduction()` which covers full reductions, reductions over the jagged dim, and reductions over the non-jagged dim
    * **TODO: Handle other types of OpInfos manually by writing sample_inputs_funcs!**
* Several xfails were added to get things passing

TODO:
* Pass non-contiguous / non-contiguous with holes NJTs

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) to eager
    * `test_backward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `BinaryUfuncInfo`s utilize a new `sample_inputs_binary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
        * **TODO: mixed (NT, T), (T, NT) inputs**
    * `ReductionOpInfo`s utilize a new `sample_inputs_njt_reduction()` which covers full reductions, reductions over the jagged dim, and reductions over the non-jagged dim
    * **TODO: Handle other types of OpInfos manually by writing sample_inputs_funcs!**
* Several xfails were added to get things passing

TODO:
* Pass non-contiguous / non-contiguous with holes NJTs

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) to eager
    * `test_backward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `BinaryUfuncInfo`s utilize a new `sample_inputs_binary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
        * **TODO: mixed (NT, T), (T, NT) inputs**
    * `ReductionOpInfo`s utilize a new `sample_inputs_njt_reduction()` which covers full reductions, reductions over the jagged dim, and reductions over the non-jagged dim
    * **TODO: Handle other types of OpInfos manually by writing sample_inputs_funcs!**
* Several xfails were added to get things passing

TODO:
* Pass non-contiguous / non-contiguous with holes NJTs

[ghstack-poisoned]

jbschlosser committed:
Update on "Initial OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) to eager
    * `test_backward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `BinaryUfuncInfo`s utilize a new `sample_inputs_binary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `ReductionOpInfo`s utilize a new `sample_inputs_njt_reduction()` which covers full reductions, reductions over the jagged dim, and reductions over the non-jagged dim
* Several xfails were added to get things passing

TODO (future PRs):
* Pass non-contiguous / non-contiguous with holes NJTs (maybe we should have separate tests for these? most ops don't support NJTs with holes today)
* Mixed (NT, T), (T, NT) inputs for binary ops
* Handle other types of OpInfos (beyond unary pointwise, binary pointwise, and reduction) by manually by writing sample_inputs_funcs
* Address all xfails via fixes

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/opinfo/definitions/nested.py:
```diff
@@ -0,0 +1,277 @@
+    new_op.supports_njt = True
+
+    if op.full_name in njt_sample_inputs:
+        new_op.sample_inputs_func = njt_sample_inputs[op.full_name]
```
One reason to have a separate `nt_sample_inputs_func` rather than reuse the original `sample_inputs_func` field is if the person writing the sample inputs wants to do some transformation on the original inputs to produce the nested ones. 

But I'm also sure how practical automated transforms are on the original inputs compared to just writing test cases manually:
- For the simple case, where the nested tensor logic is unaffected by the non-tensor args, the bit of extra coverage that gives is that we'd be testing that the non-tensor args are properly propagated to the desugared call.  In this case, manually writing a few inputs to spot-check should be enough.
- For the case where nested tensor logic must take into account the argument, e.g. "dim" or "dims" its possible that at a certain point, it may just be more practical to manually write the sample inputs - e.g. for view ops.

jbschlosser replied:
> One reason to have a separate nt_sample_inputs_func rather than reuse the original sample_inputs_func field is if the person writing the sample inputs wants to do some transformation on the original inputs to produce the nested ones.

I think this would still be possible with the current structure - you'd basically write a transformation func that utilizes the original `sample_inputs_func` to produce the new one. IIUC it's not strictly necessary to have a separate `nt_sample_inputs_func` field to accomplish this.

> But I'm also not sure how practical automated transforms are on the original inputs compared to just writing test cases manually:
> For the simple case, where the nested tensor logic is unaffected by the non-tensor args, the bit of extra coverage that gives is that we'd be testing that the non-tensor args are properly propagated to the desugared call. In this case, manually writing a few inputs to spot-check should be enough.

True, and maybe it's worth revisiting because it's nice that we can generally expect existing `OpInfo`s to cover non-tensor args pretty extensively. I don't think anything precludes us from adding this sorta thing on in the future.

> For the case where nested tensor logic must take into account the argument, e.g. "dim" or "dims" its possible that at a certain point, it may just be more practical to manually write the sample inputs - e.g. for view ops.
 
I'm glad you bring up view ops - I've been thinking of some view-specific testing we could do (it's certainly useful to have different views as input to a torch.compiled function to test out view fake-ification; we could move those tests over from `test/dynamo/test_subclasses.py`). I do agree that it's probably more practical to manually write the view sample inputs to ensure we're covering everything we want to, but there's still some view-specific testing we could write generally. There's always a tradeoff between generality and exhaustive coverage, and I think we want a little of both general and specific tests here.

soulitzer replied:
> I think this would still be possible with the current structure - you'd basically write a transformation func that utilizes the original sample_inputs_func to produce the new one. IIUC it's not strictly necessary to have a separate nt_sample_inputs_func field to accomplish this.

Good point, though I can imagine it being easier to write sample inputs function that operate generically over a set of operators. 

> There's always a tradeoff between generality and exhaustive coverage, and I think we want a little of both general and specific tests here.

Agreed. Definitely good to have a good baseline of testing through the OpInfo in addition to specific tests.

soulitzer commented on torch/testing/_internal/opinfo/definitions/nested.py:
```diff
@@ -0,0 +1,277 @@
+    ):
+        # TODO: account for non-contiguous NJTs here
+        njt2 = torch.randn_like(njt1)
+        yield SampleInput(njt1, args=(njt2,), kwargs=dict(op_kwargs))
```
Could add a TODO to test the various broadcasting cases.

soulitzer commented on torch/testing/_internal/opinfo/definitions/nested.py:
```diff
@@ -0,0 +1,277 @@
+    for njt in _sample_njts(
+        device=device, dtype=dtype, requires_grad=requires_grad, dims=[2, 3, 4]
+    ):
+        yield SampleInput(njt)
```
I can imagine this being supplemented with one or two test cases that test that non-tensor args are properly propagated

soulitzer reviewed (approved):
Looking pretty good! Added some minor comments. I think we should land this sooner than later, and future work if necessary can be parallelized more.

jbschlosser committed:
Update on "Initial OpInfo-based testing for NJTs"


This PR utilizes the info from the existing OpInfo database `op_db` to contribute to general NJT testing.
* New tests in `TestNestedTensorOpInfo`
    * `test_forward()` - compares forward output to an unbind-based reference
    * `test_backward()` - compares forward output and grads to an unbind-based reference
    * `test_forward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) to eager
    * `test_backward_compile()` - compares forward compile output (`backend="aot_eager_decomp_partition"`) and grads to eager
* To avoid adding a bunch of NJT-specific stuff to the `OpInfo` structure, this PR translates `op_db` -> a NJT-specific `njt_op_db`.
    * `UnaryUfuncInfo`s utilize a new `sample_inputs_unary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `BinaryUfuncInfo`s utilize a new `sample_inputs_binary_njt_pointwise()` which iterates through a comprehensive list of NJTs: contiguous / non-contiguous, dims 2, 3, and 4, transposed / not, etc.
    * `ReductionOpInfo`s utilize a new `sample_inputs_njt_reduction()` which covers full reductions, reductions over the jagged dim, and reductions over the non-jagged dim
* Several xfails were added to get things passing

TODO (future PRs):
* Pass non-contiguous / non-contiguous with holes NJTs (maybe we should have separate tests for these? most ops don't support NJTs with holes today)
* Mixed (NT, T), (T, NT) inputs for binary ops
* Handle other types of OpInfos (beyond unary pointwise, binary pointwise, and reduction) by manually by writing sample_inputs_funcs
* Address all xfails via fixes

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
